apps/arena/native/physics/src/lib.rs

```diff
@@ -77,20 +77,50 @@ fn move_entity_to_position(
     entity: Entity,
     new_position: Position,
     external_wall: Entity,
+    obstacles: HashMap<u64, Entity>,
 ) -> Entity {
     let mut entity: Entity = entity;
     entity.position = new_position;
 
     if entity.category == Category::Player && !entity.is_inside_map(&external_wall) {
         entity.move_to_next_valid_position_inside(&external_wall);
     }
+
+    let collides_with = entity.collides_with(obstacles.clone().into_values().collect());
+
+    if entity.category == Category::Player && !collides_with.is_empty() {
+        let collided_with: Vec<&Entity> = collides_with
+            .iter()
+            .map(|id| obstacles.get(id).unwrap())
+            .collect();
+        entity.move_to_next_valid_position_outside(collided_with);
+    }
     entity
 }
 
 #[rustler::nif()]
-fn move_entity_to_direction(entity: Entity, direction: Position, amount: f32) -> Entity {
+fn move_entity_to_direction(
+    entity: Entity,
+    direction: Position,
+    amount: f32,
+    external_wall: Entity,
+    obstacles: HashMap<u64, Entity>,
+) -> Entity {
     let mut entity: Entity = entity;
     entity.move_entity_to_direction(direction, amount);
+    if entity.category == Category::Player && !entity.is_inside_map(&external_wall) {
+        entity.move_to_next_valid_position_inside(&external_wall);
+    }
+
+    let collides_with = entity.collides_with(obstacles.clone().into_values().collect());
+
+    if entity.category == Category::Player && !collides_with.is_empty() {
+        let collided_with: Vec<&Entity> = collides_with
+            .iter()
+            .map(|id| obstacles.get(id).unwrap())
+            .collect();
+        entity.move_to_next_valid_position_outside(collided_with);
+    }
```

Comment:
This is done twice, can we move it to a separate function?

Reply:
this is not just done twice but four times, i moved all this repeated code to a separated function